third_party/nvfuser/csrc/lower_loop_rotation.cpp

```diff
@@ -0,0 +1,392 @@
+  // This map is initialized with the selection from scheduler, then expanded by
+  // expandSelection.
+  std::unordered_map<IterDomain*, std::unordered_set<Statement*>>
+      to_rotate_map_;
+
+  RotateLoop(const Fusion::LoopRotationParam& params) {
+    for (auto item : params) {
+      auto concrete_id = GpuLower::current()->caMap()->getConcreteMappedID(
+          std::get<0>(item)->axis(std::get<1>(item)), IdMappingMode::LOOP);
+      to_rotate_map_[concrete_id] = std::get<2>(item);
+    }
+  }
+
+  // We use the following strategy on expr selection:
+  // - If a Val is selected, then its allocation is automatically selected.
+  // - If all the exprs in a container(ForLoop and IfThenElse) are selected,
+  //   then the container is automatically selected.
+  // This function modifies to_rotate_map_ to implement this strategy
+  void expandSelection(Expr* expr) {
+    for (auto fl : for_loops_) {
+      auto it = to_rotate_map_.find(fl->iter_domain());
+      if (it == to_rotate_map_.end()) {
+        continue;
+      }
+      auto& selection = it->second;
+      auto entireScopeSelected = [&selection](const kir::Scope& scope) {
+        return std::all_of(
+            scope.exprs().begin(),
+            scope.exprs().end(),
+            [&selection](Expr* expr) { return selection.count(expr) > 0; });
+      };
+      bool should_select_this = false;
+      if (auto fl = dynamic_cast<kir::ForLoop*>(expr)) {
+        should_select_this = !fl->empty() && entireScopeSelected(fl->body());
+      } else if (auto ite = dynamic_cast<kir::IfThenElse*>(expr)) {
+        should_select_this = !ite->empty() &&
+            entireScopeSelected(ite->thenBody()) &&
+            entireScopeSelected(ite->elseBody());
+      } else if (auto alloc = dynamic_cast<kir::Allocate*>(expr)) {
+        should_select_this = selection.count(alloc->buffer()) > 0;
+      } else {
+        should_select_this = std::any_of(
+            expr->outputs().begin(),
+            expr->outputs().end(),
+            [&selection](Val* v) { return selection.count(v) > 0; });
+      }
+      if (should_select_this) {
+        selection.insert(expr);
+      }
+    }
+  }
+
+  // Expressions selected in to_rotate_map_ is not necessarily on the top of the
+  // loop, for example, if I have a loop
+  //   for (int i = 0; i < n; i++) {
+  //     i1 = i + 1;
+  //     i2 = i1 + 1;
+  //     i3 = i % 3;
+  //     i4 = i3 + 1;
+  //     i5 = i4 + 1;
+  //   }
+  // Then a valid selection is {i3, i4}, and the rotated loop will look like
+  //   if (0 < n) {
+  //     i3 = 0 % 3;
+  //     i4 = i3 + 1;
+  //   }
+  //   for (int i = 0; i < n; i++) {
+  //     i1 = i + 1;
+  //     i2 = i1 + 1;
+  //     i5 = i4 + 1;
+  //     if (i + 1 < n) {
+  //       i3 = (i + 1) % 3;
+  //       i4 = i3 + 1;
+  //     }
+  //   }
+  // This function validates that if the selection comply with data dependency.
+  //
+  // Note that right now, we don't have a complete data dependency analysis, so
```

Comment:
So, does this mean that even if a proper set of tensors are selected as rotated tensors when scheduling the fusion, the rotation could fail depending on how the expressions are ordered, and the failure doesn't cause the lowering to fail. 

Should this be the default behavior? Wouldn't you want the rotation happens whenever it's requested, and the lowering should fail if not? Assuming this optimization is only used for matmuls, shouldn't we prefer enforcing the rotation rather than gracefully ignoring the failure and silently causing slowdowns?

Reply:
Yeah, this makes sense. Changed.